File: 2025/day_02/s.py
from pathlib import Path

example = Path(__file__).parent / "example.txt"
true_input = Path(__file__).parent / "input.txt"
to_read = true_input

print(f"running with {to_read.name}")
data = open(to_read).read().strip()


def invalid(n: int):
    ns = str(n)
    for i in range(len(ns)):
        if ns[:i] == ns[i:]:
            return True
    return False


def invalidp2(n: int):
    ns = str(n)
    for i in range(1, len(ns)):
        if len(ns) % i != 0:
            continue

        amount = len(ns) // i

        if all(ns[:i] == ns[rep * i : (rep + 1) * i] for rep in range(amount)):
            return True
    return False


def part1(data: list[str]):
    ranges = [c.strip().split("-") for c in data.split(",")]

    l = []
    for start, end in ranges:
        startn = int(start)
        endn = int(end)
        for n in range(startn, endn + 1):
            if invalid(n):
                l.append(n)
    return sum(l)


def part2(data: list[str]):
    ranges = [c.strip().split("-") for c in data.split(",")]

    l = []
    for start, end in ranges:
        startn = int(start)
        endn = int(end)
        for n in range(startn, endn + 1):
            if invalidp2(n):
                l.append(n)
    return sum(l)


print(part1(data))
print(part2(data))

File: 2025/day_02/s.hs
splitOn :: Char -> String -> [String]
splitOn sep "" = []
splitOn sep s = takeWhile (/= sep) s : splitOn sep stuff
  where
    rest = dropWhile (/= sep) s
    stuff
      | null rest = ""
      | length rest == 1 = rest
      | otherwise = tail rest

invalidp1 :: Int -> Bool
invalidp1 n =
  let ns = show n; halfLen = length ns `div` 2
   in take halfLen ns == drop halfLen ns

rangeToInvalids :: (Int -> Bool) -> [Int] -> [Int]
rangeToInvalids invFun [low, hi] = filter invFun [low .. hi]

part1 :: [[Int]] -> Int
part1 ranges = let invalids = map (sum . rangeToInvalids invalidp1) ranges in sum invalids

patternInNum :: String -> String -> Bool
patternInNum subset ns
  | subLength > length ns = False
  | subset == ns = True
  | otherwise = (subset == take subLength ns) && patternInNum subset (drop subLength ns)
  where
    subLength = length subset

invalidp2 :: Int -> Bool
invalidp2 n =
  let ns = show n; lengthes = [1 .. length ns `div` 2]
   in or [patternInNum (take l ns) ns | l <- lengthes]

part2 :: [[Int]] -> Int
part2 ranges = let invalids = map (sum . rangeToInvalids invalidp2) ranges in sum invalids

main :: IO ()
main = do
  input <- head . lines <$> readFile "example.txt"
  let rangesStr = splitOn ',' input
  let ranges = map (map (\x -> read x :: Int) . splitOn '-') rangesStr

  print (part1 ranges)
  print (part2 ranges)
